src/bayes/sampling.jl

```diff
@@ -0,0 +1,121 @@
+"""
+    function metropolis_sample(
+        initial_tree::FelNode,
+        models::Vector{<:BranchModel},
+        num_of_samples;
+        bl_modifier::UnivariateSampler = BranchlengthSampler(Normal(0,2), Normal(-1,1))
+        burn_in=1000, 
+        sample_interval=10,
+        collect_LLs = false,
+        midpoint_rooting=false,
+    )
+
+Samples tree topologies from a posterior distribution. felsenstein! should be called on the initial tree before calling this function.
+
+# Arguments
+- `initial_tree`: An initial topology with (important!) the leaves populated with data, for the likelihood calculation.
+- `models`: A list of branch models.
+- `num_of_samples`: The number of tree samples drawn from the posterior.
+- `bl_sampler`: Sampler used to drawn branchlengths from the posterior. 
+- `burn_in`: The number of samples discarded at the start of the Markov Chain.
+- `sample_interval`: The distance between samples in the underlying Markov Chain (to reduce sample correlation).
+- `collect_LLs`: Specifies if the function should return the log-likelihoods of the trees.
+- `midpoint_rooting`: Specifies whether the drawn samples should be midpoint rerooted (Important! Should only be used for time-reversible branch models starting in equilibrium).
+
+# Returns
+- `samples`: The trees drawn from the posterior.
+- `sample_LLs`: The associated log-likelihoods of the tree (optional).
+"""
+function metropolis_sample(
+    initial_tree::FelNode,
+    models::Vector{<:BranchModel},
+    num_of_samples;
+    bl_sampler::UnivariateSampler = BranchlengthSampler(Normal(0,2), Normal(-1,1)),
+    burn_in=1000, 
+    sample_interval=10,
+    collect_LLs = false,
+    midpoint_rooting=false,
+    ladderize = false,
+    )
+
+    # The prior over the (log) of the branchlengths should be specified in bl_sampler. 
+    # Furthermore, a non-informative/uniform prior is assumed over the tree topolgies (excluding the branchlengths).
+
+    sample_LLs = []
+    samples = FelNode[]
+    tree = deepcopy(initial_tree)
+    iterations = burn_in + num_of_samples * sample_interval
+    
+    softmax_sampler = x -> rand(Categorical(softmax(x)))
```

Comment:
As of right now, the name of the function contains "optim" so defaulting to something that is not optimising might be confusing, therefore I think should default to argmax. At the same time, as the functionality of nni_optim! is extended in this PR, the function name nni_optim! (and this also applies for nni_branchlength!) should maybe be reconsidered (since we are already breaking backward compatibility), and then it might be a sensible thing to default to softmax. Some name suggestions might be nni_operation!, nni_action! etc.

Reply:
My comment was directed to the `metropolis_sample` function.

I've also been considering renaming `nni/branchlength_optim!` but haven't come up with a good substitute.